rclcpp/test/test_publisher_subscription_count_api.cpp

```diff
@@ -0,0 +1,142 @@
+// limitations under the License.
+
+#include <gtest/gtest.h>
+
+#include <string>
+#include <memory>
+
+#include "rclcpp/exceptions.hpp"
+#include "rclcpp/publisher.hpp"
+#include "rclcpp/rclcpp.hpp"
+
+#include "rcl_interfaces/msg/intra_process_message.hpp"
+
+using rcl_interfaces::msg::IntraProcessMessage;
+
+class TestPublisherSubscriptionCount : public ::testing::Test
+{
+protected:
+  static void SetUpTestCase()
+  {
+    rclcpp::init(0, nullptr);
+  }
+
+  void SetUp()
+  {
+    node = std::make_shared<rclcpp::Node>(
+      "my_node",
+      "/ns",
+      rclcpp::NodeOptions().use_intra_process_comms(true));
+    publisher = node->create_publisher<IntraProcessMessage>("/topic");
+  }
+
+  void TearDown()
+  {
+    node.reset();
+  }
+
+  void test_common(rclcpp::NodeOptions options, const uint64_t intraprocess_count_results[2]);
+
+  rclcpp::Node::SharedPtr node;
+  std::shared_ptr<rclcpp::PublisherBase> publisher;
+  static std::chrono::milliseconds offset;
+};
+
+std::chrono::milliseconds TestPublisherSubscriptionCount::offset = std::chrono::milliseconds(2000);
+
+void OnMessage(const rcl_interfaces::msg::IntraProcessMessage::SharedPtr msg)
+{
+  (void)msg;
+}
+
+void TestPublisherSubscriptionCount::test_common(
+  rclcpp::NodeOptions options,
+  const uint64_t intraprocess_count_results[2])
+{
+  EXPECT_EQ(publisher->get_subscription_count(), 0u);
+  EXPECT_EQ(publisher->get_intra_process_subscription_count(), 0u);
+  {
+    auto sub = node->create_subscription<IntraProcessMessage>("/topic", &OnMessage);
+    rclcpp::sleep_for(offset);
+    EXPECT_EQ(publisher->get_subscription_count(), 1u);
+    EXPECT_EQ(publisher->get_intra_process_subscription_count(), intraprocess_count_results[0]);
+    {
+      rclcpp::Node::SharedPtr another_node = std::make_shared<rclcpp::Node>(
+        "another_node",
+        "/ns",
+        options);
+      auto another_sub =
+        another_node->create_subscription<IntraProcessMessage>("/topic", &OnMessage);
+
+      rclcpp::sleep_for(offset);
+      EXPECT_EQ(publisher->get_subscription_count(), 2u);
+      EXPECT_EQ(publisher->get_intra_process_subscription_count(), intraprocess_count_results[1]);
+    }
+    rclcpp::sleep_for(offset);
+    EXPECT_EQ(publisher->get_subscription_count(), 1u);
+    EXPECT_EQ(publisher->get_intra_process_subscription_count(), intraprocess_count_results[0]);
+  }
+  /**
+    * Counts should be zero here, as all are subscriptions are out of scope.
+    * Subscriptions count checking is always preceeded with an sleep, as random failures had been
+    * detected without it. */
+  rclcpp::sleep_for(offset);
+  EXPECT_EQ(publisher->get_subscription_count(), 0u);
+  EXPECT_EQ(publisher->get_intra_process_subscription_count(), 0u);
+}
+
+/*
+   Testing publisher subscription count api and internal process subscription count.
+   Two subscriptions in same using intra-process comm.
+ */
+TEST_F(TestPublisherSubscriptionCount, test_two_intra_comm) {
```

Comment:
@ivanpauno consider `gtest` parameterized tests instead of re-implementing their functionality.